src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,262 @@
+            }
+
+            return strings;
+        }
+
+        internal IReadOnlyList<string> ReadDelimitedString()
+        {
+            if (Read())
+            {
+                switch (TokenType)
+                {
+                    case JsonTokenType.String:
+                        var value = GetString();
+
+                        return value.Split(DelimitedStringDelimiters, StringSplitOptions.RemoveEmptyEntries);
+
+                    default:
+                        var invalidCastException = new InvalidCastException();
+                        throw new JsonException(invalidCastException.Message, invalidCastException);
+                }
+            }
+
+            return null;
+        }
+
+        internal List<string> ReadStringArrayAsList(List<string> strings = null)
+        {
+            if (TokenType == JsonTokenType.StartArray)
+            {
+                while (Read() && TokenType != JsonTokenType.EndArray)
+                {
+                    string value = _reader.ReadTokenAsString();
+
+                    strings = strings ?? new List<string>();
+
+                    strings.Add(value);
+                }
+            }
+
+            return strings;
+        }
+
+        internal bool ReadNextTokenAsBoolOrFalse()
+        {
+            if (Read() && (TokenType == JsonTokenType.False || TokenType == JsonTokenType.True))
+            {
+                return GetBoolean();
+            }
+            return false;
+        }
+
+        internal IReadOnlyList<string> ReadNextStringOrArrayOfStringsAsReadOnlyList()
+        {
+            if (Read())
+            {
+                switch (_reader.TokenType)
+                {
+                    case JsonTokenType.String:
+                        return new[] { (string)_reader.GetString() };
+
+                    case JsonTokenType.StartArray:
+                        return ReadStringArrayAsReadOnlyListFromArrayStart();
+
+                    case JsonTokenType.StartObject:
+                        return null;
+                }
+            }
+
+            return null;
+        }
+
+        internal IReadOnlyList<string> ReadStringArrayAsReadOnlyListFromArrayStart()
+        {
+            List<string> strings = null;
+
+            while (Read() && _reader.TokenType != JsonTokenType.EndArray)
+            {
+                string value = _reader.ReadTokenAsString();
+
+                strings = strings ?? new List<string>();
+
+                strings.Add(value);
+            }
+
+            return (IReadOnlyList<string>)strings ?? Array.Empty<string>();
+        }
+
+        private void GetMoreBytesFromStream()
+        {
+            if (_reader.BytesConsumed < _buffer.Length)
+            {
+                ReadOnlySpan<byte> leftover = _buffer.AsSpan((int)_reader.BytesConsumed);
+                if (leftover.Length == _buffer.Length)
+                {
+                    Array.Resize(ref _buffer, _buffer.Length * 2);
```

Comment:
[The docs for `Array.Resize`](https://learn.microsoft.com/en-us/dotnet/api/system.array.resize?view=net-8.0#system-array-resize-1(-0()@-system-int32)) says:

> This method allocates a new array with the specified size, copies elements from the old array to the new one, and then replaces the old array with the new one.

If most of the bytes from the buffer are consumed already, there's no point having `Array.Resize` copy them.

```suggestion
                    _buffer = new byte[_buffer.Length * 2];
```


But there is also the array pool question from elsewhere.